src/misc/cycle_bases.c

```diff
@@ -0,0 +1,507 @@
+ *   may not be of the smallest possible size.
+ *   \p cutoff is used to limit the depth of the BFS tree when computing candidate
+ *   cycles. Specifying a cutoff can speed up the computation substantially.
+ * \param complete Boolean value. Used only when \p cutoff was given.
+ *   If true, a complete basis is returned. If false, only cycles not greater
+ *   than <code>2*cutoff + 1</code> are returned. This may save computation
+ *   time, however, the result will not span the entire cycle space.
+ * \param use_cycle_order If true, each cycle is returned in natural order:
+ *   the edge IDs will appear ordered along the cycle. This comes at a small
+ *   performance cost. If false, no guarantees are given about the ordering
+ *   of edge IDs within cycles. This parameter exists solely to control
+ *   performance tradeoffs
+ * \param result An initialized integer vector list, the elements of the cycle
+ *   basis will be stored here as vectors of edge IDs.
+ * \return Error code.
+ */
+igraph_error_t igraph_minimum_cycle_basis(const igraph_t *graph,
+                                          igraph_integer_t cutoff,
+                                          igraph_bool_t complete,
+                                          igraph_bool_t use_cycle_order,
+                                          igraph_vector_int_list_t *result) {
+
+    igraph_integer_t no_of_nodes = igraph_vcount(graph);
+    igraph_integer_t no_of_edges = igraph_ecount(graph);
+    igraph_integer_t i;
+    igraph_integer_t rank;
+    igraph_vector_int_list_t candidates;
+
+    /* Compute candidate elements for the minimum weight basis. */
+    {
+        igraph_inclist_t inclist;
+        igraph_vector_int_t visited;
+        igraph_vector_int_t degrees;
+        igraph_integer_t no_of_comps;
+        igraph_integer_t mark;
+
+        /* We use the degrees to avoid doing a BFS from vertices with d < 3, except in special cases.
+         * Degrees cannot be computed from the inclist because there we use IGRAPH_LOOPS_ONCE. */
+        IGRAPH_VECTOR_INT_INIT_FINALLY(&degrees, no_of_nodes);
+        IGRAPH_CHECK(igraph_degree(graph, &degrees, igraph_vss_all(), IGRAPH_ALL, IGRAPH_LOOPS));
+
+        IGRAPH_CHECK(igraph_connected_components(graph, NULL, NULL, &no_of_comps, IGRAPH_WEAK));
+        rank = no_of_edges - no_of_nodes + no_of_comps;
+
+        IGRAPH_VECTOR_INT_INIT_FINALLY(&visited, no_of_nodes);
+
+        IGRAPH_CHECK(igraph_inclist_init(graph, &inclist, IGRAPH_ALL, IGRAPH_LOOPS_ONCE));
+        IGRAPH_FINALLY(igraph_inclist_destroy, &inclist);
+
+         /* TODO: estimate space to reserve. 'rank' is a lower bound only. */
+        IGRAPH_VECTOR_INT_LIST_INIT_FINALLY(&candidates, 0);
+        IGRAPH_CHECK(igraph_vector_int_list_reserve(&candidates, rank));
+
+        mark = 0;
+        for (i=0; i < no_of_nodes; ++i) {
+            igraph_integer_t degree = VECTOR(degrees)[i];
+            igraph_bool_t vis = VECTOR(visited)[i] % 3;
```

Comment:
Explain where the `% 3` comes from. I understand, but it's better to be explicit, especially because we are talking about "vertices of degree 3 or greater" one line below, and this 3 has nothing to do with that 3.

Also note that with this trick, we are limiting ourselves to processing graphs where `no_of_nodes <= IGRAPH_INTEGER_MAX / 3`, otherwise the markers start to wrap around in `visited`. It's not a problem for me, just let us be aware of it.

Reply:
Note to self: implement overflow checks.